ares/n64/cpu/memory.cpp

```diff
@@ -168,132 +169,40 @@ inline auto CPU::busReadBurst(u32 address, u32 *data) -> void {
   return bus.readBurst<Size>(address, data, *this);
 }
 
-auto CPU::fetch(u64 vaddr) -> maybe<u32> {
-  if(vaddrAlignedError<Word>(vaddr, false)) return nothing;
-  switch(segment(vaddr)) {
-  case Context::Segment::Unused:
-    step(1 * 2);
-    addressException(vaddr);
-    exception.addressLoad();
-    return nothing;
-  case Context::Segment::Mapped:
-    if(auto match = tlb.load(vaddr)) {
-      if(match.cache) return icache.fetch(vaddr, match.paddr & context.physMask, cpu);
-      step(1 * 2);
-      return busRead<Word>(match.paddr & context.physMask);
-    }
-    step(1 * 2);
-    addressException(vaddr);
-    return nothing;
-  case Context::Segment::Cached:
-    return icache.fetch(vaddr, vaddr & 0x1fff'ffff, cpu);
-  case Context::Segment::Cached32:
-    return icache.fetch(vaddr, vaddr & 0xffff'ffff, cpu);
-  case Context::Segment::Direct:
-    step(1 * 2);
-    return busRead<Word>(vaddr & 0x1fff'ffff);
-  case Context::Segment::Direct32:
-    step(1 * 2);
-    return busRead<Word>(vaddr & 0xffff'ffff);
-  }
-
-  unreachable;
+auto CPU::fetch(PhysAccess access) -> maybe<u32> {
+  step(1 * 2);
```

Comment:
This doesn't appear to preserve the original stepping behavior. The original code deferred to InstructionCache::fetch to step the appropriate amount (in cases where it is called). Now we will end up with an extra step.

Reply:
True, I'll remove this then.

Reply:
Removing it entirely means the non-cached cases that used to have a step now have none. Does the rationale in your reply on the CPU::read() comment thread (justifying the removal) apply here as well?

Reply:
Unfortunately this becomes a giant snowball pretty fast. If you can suggest a simple way to maintain the current timing behavior in the PR I can do that, otherwise we will have to accept small changes. The timing is wrong anyway. The concept that the cache has an impact on timing is wrong. Timing on the CPU is either stalls or the bus being held during memory accesses (and depending on what access), cache has nothing to do with it. I tried fixing it more but I end up needing to implement more correct timings across the board, and I don't have numbers for those, and this in turn has surprising behaviors on random games.

Reply:
I pushed a commit that should get us closer to before, let me know if it works for you

Reply:
Yeah, that seems less disruptive.